torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1305,7 +1305,18 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                 if self.gradient_postdivide_factor > 1:
                     # Average grad by world_size for consistency with PyTorch DDP.
                     output.div_(self.gradient_postdivide_factor)
-                param.grad.data = output
+                accumulate_grad = getattr(param, "_saved_grad_shard", None) is not None
+                if accumulate_grad:
+                    p_assert(
+                        param._saved_grad_shard.shape == output.shape,  # type: ignore[attr-defined]
+                        "Shape mismatch when accumulating gradients: "  # type: ignore[attr-defined]
+                        f"existing shape={param._saved_grad_shard.shape} "
+                        f"new shape={output.shape}"  # type: ignore[attr-defined]
+                    )
+                    param._saved_grad_shard.data += output.data  # type: ignore[attr-defined]
+                else:
+                    param._saved_grad_shard = output.data  # type: ignore[attr-defined]
```

Comment:
are we switching from using `output` to `output.data` in non-grad accumulation use case?

Reply:
Good question. This is good for me to clarify. Was there any reason to use `output` before?

In my understanding, it does not matter since we are not supporting taking the gradient of the gradient, so we do not need to record any operations on `output` in the autograd graph. Fairscale uses `output.data` in both code paths, though I do not see why either way is better for the non-gradient accumulation case.